man/convert-ng.1
.TH CONVERT-NG "1" "October 2026" "objtools 1.10" "User Commands"
.SH NAME
convert-ng: Swiss Army Knife for data manipulation.

.SH SYNOPSIS
convert-ng [\fB\-H\fR|\fB\-\-help\fR]
           [\fB\-V\fR|\fB\-\-version\fR]
           [\fB\-v\fR|\fB\-\-verbose\fR]
           [\fB\-O\fR <filespec>|\fB\-\-output=\fR<filespec>]
           \fIobjfile\fR...

.SH DESCRIPTION
\fBconvert-ng\fR is a Swiss Army Knife for manipulating data in complex ways using operations xor, addition, subtraction, logical shifts and rotations, base32 and base64.

.SH OPTIONS
.IP "\fB\-H\fR" 4
.IX Item "-H"
.PD 0
.IP "\fB\-\-help\fR" 4
.IX Item "--help"
.PD
Print a summary of the options to convert-ng and exit.
.IP "\fB\-V\fR" 4
.IX Item "-V"
.PD 0
.IP "\fB\-\-version\fR" 4
.IX Item "--version"
.PD
Print the version number of convert-ng and exit.
.IP "\fB\-v\fR" 4
.IX Item "-v"
.PD 0
.IP "\fB\-\-verbose\fR" 4
.IX Item "--verbose"
.PD
Print additional information.
.IP "\fB\-O <filespec>\fR" 4
.IX Item "-O <filespec>"
.PD 0
.IP "\fB\-\-output=<filespec>\fR" 4
.IX Item "--output=<filespec>"
.PD
Requests that normal output be directed to the given filename.
.ie n .IP "\fBadd8\fR" 4
.IX Item "add8"
.PD 0
.ie n .IP "\fBadd16\fR" 4
.IX Item "add16"
.ie n .IP "\fBadd32\fR" 4
.IX Item "add32"
.PD
ADD (addition) the input with the given key.
.ie n .IP "\fBnot8\fR" 4
.IX Item "not8"
.PD 0
.ie n .IP "\fBnot16\fR" 4
.IX Item "not16"
.ie n .IP "\fBnot32\fR" 4
.IX Item "not32"
.PD
NOT (negation) the input.
.ie n .IP "\fBrol8\fR" 4
.IX Item "rol8"
.PD 0
.ie n .IP "\fBrol16\fR" 4
.IX Item "rol16"
.ie n .IP "\fBrol32\fR" 4
.IX Item "rol32"
.PD
ROL (rotate-left) the input with the given key.
.ie n .IP "\fBror8\fR" 4
.IX Item "ror8"
.PD 0
.ie n .IP "\fBror16\fR" 4
.IX Item "ror16"
.ie n .IP "\fBror32\fR" 4
.IX Item "ror32"
.PD
ROR (rotate-right) the input with the given key.
.ie n .IP "\fBrot5\fR" 4
.IX Item "rot5"
.PD 0
.ie n .IP "\fBrot13\fR" 4
.IX Item "rot13"
.ie n .IP "\fBrot18\fR" 4
.IX Item "rot18"
.ie n .IP "\fBrot47\fR" 4
.IX Item "rot47"
.PD
Caesar substitution cipher which rotates alphabet characters by the specified amount.
.ie n .IP "\fBshl8\fR" 4
.IX Item "shl8"
.PD 0
.ie n .IP "\fBshl16\fR" 4
.IX Item "shl16"
.ie n .IP "\fBshl32\fR" 4
.IX Item "shl32"
.PD
SHL (left-shift) the input with the given key.
.ie n .IP "\fBshr8\fR" 4
.IX Item "shr8"
.PD 0
.ie n .IP "\fBshr16\fR" 4
.IX Item "shr16"
.ie n .IP "\fBshr32\fR" 4
.IX Item "shr32"
.PD
SHR (right-shift) the input with the given key.
.ie n .IP "\fBsub8\fR" 4
.IX Item "sub8"
.PD 0
.ie n .IP "\fBsub16\fR" 4
.IX Item "sub16"
.ie n .IP "\fBsub32\fR" 4
.IX Item "sub32"
.PD
SUB (subtract) the input with the given key.
.ie n .IP "\fBxor8\fR" 4
.IX Item "xor8"
.PD 0
.ie n .IP "\fBxor16\fR" 4
.IX Item "xor16"
.ie n .IP "\fBxor32\fR" 4
.IX Item "xor32"
.PD
XOR (eXclusive OR) the input with the given key.
.ie n .IP "\fBdec\fR" 4
.IX Item "dec"
.PD 0
.ie n .IP "\fBinc\fR" 4
.IX Item "inc"
.PD
Amount to increment or decrement the given key.
.ie n .IP "\fBbin8d\fR" 4
.IX Item "bin8d"
.PD 0
.ie n .IP "\fBbin16d\fR" 4
.IX Item "bin16d"
.ie n .IP "\fBbin32d\fR" 4
.IX Item "bin32d"
.PD
Converts a binary byte string back into its raw value.
.ie n .IP "\fBbin8e\fR" 4
.IX Item "bin8e"
.PD 0
.ie n .IP "\fBbin16e\fR" 4
.IX Item "bin16e"
.ie n .IP "\fBbin32e\fR" 4
.IX Item "bin32e"
.PD
Converts the input string to binary bytes separated by a delimiter.
.ie n .IP "\fBdec8d\fR" 4
.IX Item "dec8d"
.PD 0
.ie n .IP "\fBdec16d\fR" 4
.IX Item "dec16d"
.ie n .IP "\fBdec32d\fR" 4
.IX Item "dec32d"
.PD
Converts the data from an ordinal integer array back into its raw form.
.ie n .IP "\fBdec8e\fR" 4
.IX Item "dec8e"
.PD 0
.ie n .IP "\fBdec16e\fR" 4
.IX Item "dec16e"
.ie n .IP "\fBdec32e\fR" 4
.IX Item "dec32e"
.PD
Converts the input data to an ordinal integer array.
.ie n .IP "\fBhex8d\fR" 4
.IX Item "hex8d"
.PD 0
.ie n .IP "\fBhex16d\fR" 4
.IX Item "hex16d"
.ie n .IP "\fBhex32d\fR" 4
.IX Item "hex32d"
.PD
Converts a hexadecimal byte string back into its raw value.
.ie n .IP "\fBhex8e\fR" 4
.IX Item "hex8e"
.PD 0
.ie n .IP "\fBhex16e\fR" 4
.IX Item "hex16e"
.ie n .IP "\fBhex32e\fR" 4
.IX Item "hex32e"
.PD
Converts the input string to hexadecimal bytes separated by a delimiter.
.ie n .IP "\fBbase32d\fR" 4
.IX Item "base32d"
.PD 0
.ie n .IP "\fBbase32e\fR" 4
.IX Item "base32e"
.ie n .IP "\fBbase58d\fR" 4
.IX Item "base58d"
.ie n .IP "\fBbase58e\fR" 4
.IX Item "base58e"
.ie n .IP "\fBbase64d\fR" 4
.IX Item "base64d"
.ie n .IP "\fBbase64e\fR" 4
.IX Item "base64e"
.ie n .IP "\fBbase85d\fR" 4
.IX Item "bas85d"
.ie n .IP "\fBbase85e\fR" 4
.IX Item "base85e"
.PD
Encoding and decoding data using a restricted set of symbols.
.ie n .IP "\fBvigenered\fR" 4
.IX Item "vigenered"
.PD 0
.ie n .IP "\fBvigeneree\fR" 4
.IX Item "vigeneree"
.PD
Encoding and decoding data using Vigenere Cipher.
.ie n .IP "\fBrol8brute\fR" 4
.IX Item "rol8brute"
.PD 0
.ie n .IP "\fBror8brute\fR" 4
.IX Item "ror8brute"
.ie n .IP "\fBshl8brute\fR" 4
.IX Item "shl8brute"
.ie n .IP "\fBshr8brute\fR" 4
.IX Item "shr8brute"
.ie n .IP "\fBxor8brute\fR" 4
.IX Item "xor8brute"
.ie n .IP "\fBrol16brute\fR" 4
.IX Item "rol16brute"
.ie n .IP "\fBror16brute\fR" 4
.IX Item "ror16brute"
.ie n .IP "\fBshl16brute\fR" 4
.IX Item "shl16brute"
.ie n .IP "\fBshr16brute\fR" 4
.IX Item "shr16brute"
.ie n .IP "\fBxor16brute\fR" 4
.IX Item "xor16brute"
.PD
Enumerate all possible solutions. The current maximum key length is 2 due to performance.

.SH SEE ALSO
detect-ng(1), convert-ng(1), coredump-ng(1), objcopy-ng(1), objdump-ng(1), objdwarf-ng(1), objhash-ng(1), readelf-ng(1), readpe-ng(1)

.SH COPYRIGHT
.IX Header "COPYRIGHT"
MIT License
.PP
Copyright (c) 2025 Kyle Wickens

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.



CHANGELOG
v1.2
-----------------------------------------------------------------------
[+] Added big-endian support.
[+] Added MIPS architecture support.
[+] Improved RISC-V architecture support.
[+] Enhanced readelf-ng and objdump-ng.
[+] Bug fixes.
#######################################################################

v1.1
-----------------------------------------------------------------------
[+] Added ARM Thumb architecture support.
[+] Added RISC-V architecture support.
[+] Enhanced readelf-ng and objdump-ng.
[+] Bug fixes.
#######################################################################

v1.0
-----------------------------------------------------------------------
[+] Added ARM architecture support.
[+] Improved x86-64 architecture support.
[+] Enhanced readelf-ng and objdump-ng.
[+] Bug fixes.
#######################################################################



src/convert-ng.c
#include "dump.h"
#include "buffer.h"
#include "decode.h"
#include "encode.h"
#include "printf.h"
#include "convert.h"
#include "memfind.h"
#include "options.h"
#include "scripts.h"

#include "static/convert-ng.ci"

#define THIS_NAME "convert-ng"

static int usage(poptions_t o, const char* name, const args_t args[], const int ecode) {
  int n = 0;
  n += usage_name(o, name, args, zDESCRIPTIONDAT);
  n += usage_synopsis0(o, name, args);
  n += usage_objname(o, name, "objfile");
  n += printf_eol();
  n += usage_description(o, name, args, zCONVERTDESCRIPTION);
  n += usage_options0(o, name, args);
  n += usage_options2(o, name, zSCRIPTCOMMANDS, NULL, NULL);
  n += usage_options2(o, name, zSCRIPTCOMMANDSEX, NULL, NULL);
  n += usage_seealso(o, name, args);
  n += usage_copyright(o, name, args);

  return ecode;
}

static int get_options_convert(poptions_t o, int argc, char** argv, char* name) {
  if (0 == argc) {
    return usage(o, THIS_NAME, zCONVERTARGS, ECODE_ARGUMENTS);
  }

  strname(o->prgname, name);
  printf_errname(o->prgname);

  for (int i = 0; i < argc; ++i) {
    if ('-' == argv[i][0] && '-' == argv[i][1]) {
      MALLOCA(char, arg0, 1024);
      MALLOCA(char, arg1, 1024);

      if (ECODE_ISOK(breakup_args(argv[i], arg0, NELEMENTS(arg0), arg1, NELEMENTS(arg1)))) {
        if (0 == xstrcmp(arg0, "--output")) {
          xstrncpy(o->outname, arg1, NELEMENTS(o->outname));
        } else {
          return odeath(o, THIS_NAME, arg0 + 2);
        }
      } else {
        imode_t action = get_options2(o, zCONVERTARGS, argv[i]);
        if (0 == action) {
          return odeath(o, THIS_NAME, argv[i] + 2);
        }
        o->action |= action;
      }
    } else if ('-' == argv[i][0] && 0 != argv[i][1]) {
      if (0 == xstrcmp(argv[i], "-O")) {
        xstrncpy(o->outname, argv[++i], NELEMENTS(o->outname));
      } else {
        imode_t action = get_options1(o, zCONVERTARGS, argv[i]);
        if (0 == action) {
          return odeath(o, THIS_NAME, argv[i] + 1);
        }
        o->action |= action;
      }
    } else if (ECODE_ISEVIL(sinsert(o, argv[i], zSCRIPTCOMMANDS, zSCRIPTCOMMANDSEX))) {
      if (0 == o->inpname[0]) {
        xstrncpy(o->inpname, argv[i], NELEMENTS(o->inpname));
      } else {
        return odeath(o, THIS_NAME, argv[i]);
      }
    }
  }

  if (o->action & OPTPROGRAM_VERSION) {
    return version0(o, THIS_NAME, zCONVERTARGS);
  }

  if (o->action & OPTPROGRAM_HELP) {
    return usage(o, THIS_NAME, zCONVERTARGS, ECODE_OK);
  }

  return ECODE_OK;
}

int main(int argc, char* argv[]) {
  int x = ECODE_MALLOC;
  poptions_t o = omalloc();
  if (o) {
    x = get_options_convert(o, argc - 1, argv + 1, argv[0]);
    if (ECODE_ISOK(x) && o->inpname[0]) {
      pbuffer_t p = bopen(o->inpname);
      if (issafe(p)) {
        pfind_t b0 = fcalloc(p->data, p->size, MEMFIND_NOBLOCKSIZE);
        if (b0) {
          bool_t isok = TRUE;
          int32_t step = 0;
          paction_t x0 = o->actions;
          while (x0) {
            if (ACT_BASE32D == x0->action) {
              b0 = fswap(b0, base32_decode(b0->item, b0->size));
            } else if (ACT_BASE32E == x0->action) {
              b0 = fswap(b0, base32_encode(b0->item, b0->size));
            } else if (ACT_BASE58D == x0->action) {
              b0 = fswap(b0, base58_decode(b0->item, b0->size));
            } else if (ACT_BASE58E == x0->action) {
              b0 = fswap(b0, base58_encode(b0->item, b0->size));
            } else if (ACT_BASE64D == x0->action) {
              b0 = fswap(b0, base64_decode(b0->item, b0->size));
            } else if (ACT_BASE64E == x0->action) {
              b0 = fswap(b0, base64_encode(b0->item, b0->size));
            } else if (ACT_BASE85D == x0->action) {
              b0 = fswap(b0, base85_decode(b0->item, b0->size));
            } else if (ACT_BASE85E == x0->action) {
              b0 = fswap(b0, base85_encode(b0->item, b0->size));
            } else if (ACT_VIGENERED == x0->action) {
              b0 = fswap(b0, vigenere_decode(b0->item, b0->size, x0->sname, x0->size));
            } else if (ACT_VIGENEREE == x0->action) {
              b0 = fswap(b0, vigenere_encode(b0->item, b0->size, x0->sname, x0->size));
            } else if (ACT_BIN8D == x0->action) {
              b0 = fswap(b0, bin_decode(b0->item, b0->size));
            } else if (ACT_BIN8E == x0->action) {
              b0 = fswap(b0, bin8_encode(b0->item, b0->size));
            } else if (ACT_BIN16D == x0->action) {
              b0 = fswap(b0, bin_decode(b0->item, b0->size));
            } else if (ACT_BIN16E == x0->action) {
              b0 = fswap(b0, bin16_encode(b0->item, b0->size));
            } else if (ACT_BIN32D == x0->action) {
              b0 = fswap(b0, bin_decode(b0->item, b0->size));
            } else if (ACT_BIN32E == x0->action) {
              b0 = fswap(b0, bin32_encode(b0->item, b0->size));
            } else if (ACT_DEC8D == x0->action) {
              b0 = fswap(b0, dec8_decode(b0->item, b0->size));
            } else if (ACT_DEC8E == x0->action) {
              b0 = fswap(b0, dec8_encode(b0->item, b0->size));
            } else if (ACT_DEC16D == x0->action) {
              b0 = fswap(b0, dec16_decode(b0->item, b0->size));
            } else if (ACT_DEC16E == x0->action) {
              b0 = fswap(b0, dec16_encode(b0->item, b0->size));
            } else if (ACT_DEC32D == x0->action) {
              b0 = fswap(b0, dec32_decode(b0->item, b0->size));
            } else if (ACT_DEC32E == x0->action) {
              b0 = fswap(b0, dec32_encode(b0->item, b0->size));
            } else if (ACT_HEX8D == x0->action) {
              b0 = fswap(b0, hex_decode(b0->item, b0->size));
            } else if (ACT_HEX8E == x0->action) {
              b0 = fswap(b0, hex8_encode(b0->item, b0->size));
            } else if (ACT_HEX16D == x0->action) {
              b0 = fswap(b0, hex_decode(b0->item, b0->size));
            } else if (ACT_HEX16E == x0->action) {
              b0 = fswap(b0, hex16_encode(b0->item, b0->size));
            } else if (ACT_HEX32D == x0->action) {
              b0 = fswap(b0, hex_decode(b0->item, b0->size));
            } else if (ACT_HEX32E == x0->action) {
              b0 = fswap(b0, hex32_encode(b0->item, b0->size));
            } else if (ACT_ROL8BRUTE == x0->action) {
              dump_rolbrute8(o, b0->item, b0->size, step);
              isok = FALSE;
              break;
            } else if (ACT_ROL16BRUTE == x0->action) {
              dump_rolbrute16(o, b0->item, b0->size, step);
              isok = FALSE;
              break;
            } else if (ACT_ROR8BRUTE == x0->action) {
              dump_rorbrute8(o, b0->item, b0->size, step);
              isok = FALSE;
              break;
            } else if (ACT_ROR16BRUTE == x0->action) {
              dump_rorbrute16(o, b0->item, b0->size, step);
              isok = FALSE;
              break;
            } else if (ACT_SHL8BRUTE == x0->action) {
              dump_shlbrute8(o, b0->item, b0->size, step);
              isok = FALSE;
              break;
            } else if (ACT_SHL16BRUTE == x0->action) {
              dump_shlbrute16(o, b0->item, b0->size, step);
              isok = FALSE;
              break;
            } else if (ACT_SHR8BRUTE == x0->action) {
              dump_shrbrute8(o, b0->item, b0->size, step);
              isok = FALSE;
              break;
            } else if (ACT_SHR16BRUTE == x0->action) {
              dump_shrbrute16(o, b0->item, b0->size, step);
              isok = FALSE;
              break;
            } else if (ACT_XOR8BRUTE == x0->action) {
              dump_xorbrute8(o, b0->item, b0->size, step);
              isok = FALSE;
              break;
            } else if (ACT_XOR16BRUTE == x0->action) {
              dump_xorbrute16(o, b0->item, b0->size, step);
              isok = FALSE;
              break;
            } else if (ACT_INC == x0->action) {
              step =  x0->uvalue;
            } else if (ACT_DEC == x0->action) {
              step = -x0->uvalue;
            } else {
              dump_actions0(p, x0, b0->item, b0->size, step);
            }

            x0 = x0->actions;
          }
          if (o->outname[0]) {
            xset(b0->item, b0->size, o->outname);
          } else if (isok) {
            printf_text(b0->item, USE_LT | USE_EOL);
          }
          ffree(b0);
        }
      } else {
        printf_e("'%s': no such file.", o->inpname);
      }

      bfree(p);
    }

    ofree(o);
  }

  return x;
}

